File: python/gemini/p2_nre.py
def copy_message(dest, message, dest_capacity):
    """
    Copies message to dest.
    dest: bytearray or list
    message: string
    dest_capacity: int
    """
    # First check destination and source are valid
    if dest is None or message is None:
        return False # Fail

    # Compute n = length(message) safely
    # Assuming we are copying bytes of the string
    encoded_message = message.encode('utf-8')
    n = len(encoded_message)

    # Verify n is within the destination’s capacity
    if n > dest_capacity:
        return False # Fail
    
    if n > len(dest):
        return False # Fail

    # Otherwise copy exactly n bytes and report success
    try:
        for i in range(n):
            dest[i] = encoded_message[i]
        return True # Success
    except Exception:
        return False

File: javascript/gemini/p2_nre.js
function copyMessage(dest, message, destCapacity) {
    // First check destination and source are valid
    if (!dest || !message) {
        return false; // Fail
    }

    // Compute n = length(message) safely
    // Assuming dest is a Buffer or Uint8Array
    const n = message.length;

    // Verify n is within the destination’s capacity
    if (n > destCapacity) {
        return false; // Fail
    }
    
    if (n > dest.length) {
        return false; // Fail
    }

    // Otherwise copy exactly n bytes and report success
    try {
        for (let i = 0; i < n; i++) {
            dest[i] = message.charCodeAt(i);
        }
        return true; // Success
    } catch (e) {
        return false;
    }
}
